opm/output/eclipse/EclipseReader.cpp

```diff
@@ -17,230 +17,163 @@
-            } else {
-                std::string error_str = "Restart file is missing SGAS data!\n";
-                throw std::runtime_error(error_str);
-            }
+        struct keyword {
+            ecl_kw_type* kw;
+            int size;
+            const float* data;
+            const char* name;
+
+            keyword( ecl_file_type* file, const char* nm ) :
+                kw( ecl_file_iget_named_kw( file, nm, 0 ) ),
+                size( ecl_kw_get_size( kw ) ),
+                data( ecl_kw_get_float_ptr( kw ) ),
+                name( nm )
+            {}
+        };
+
+        keyword pres( file, "PRESSURE" );
+        keyword temp( file, "TEMP" );
+        keyword swat( file, "SWAT" );
+        keyword sgas( file, "SGAS" );
+
+        for( const auto& kw : { pres, temp, swat, sgas } ) {
+            if( kw.size != numcells )
+                throw std::runtime_error("Restart file: Could not restore "
+                                        + std::string( kw.name )
+                                        + ", mismatched number of cells" );
         }
-    }
-
 
-    static void restoreRSandRV(const ecl_file_type* file_type,
-                               SimulationConfigConstPtr sim_config,
-                               SimulationDataContainer& simulator_state) {
-
-        if (sim_config->hasDISGAS()) {
-            const char* RS = "RS";
-            if (ecl_file_has_kw(file_type, RS)) {
-                ecl_kw_type* rs_kw = ecl_file_iget_named_kw(file_type, RS, 0);
-                float* rs_data = ecl_kw_get_float_ptr(rs_kw);
-                auto& rs = simulator_state.getCellData( BlackoilState::GASOILRATIO );
-                for (int i = 0; i < ecl_kw_get_size( rs_kw ); i++) {
-                    rs[i] = rs_data[i];
-                }
-            } else {
-                throw std::runtime_error("Restart file is missing RS data!\n");
-            }
+        using ds = data::Solution::key;
+        data::Solution sol;
+        sol.insert( ds::PRESSURE, { pres.data, pres.data + pres.size } );
+        sol.insert( ds::TEMP,     { temp.data, temp.data + temp.size } );
+        sol.insert( ds::SWAT,     { swat.data, swat.data + swat.size } );
+        sol.insert( ds::SGAS,     { sgas.data, sgas.data + sgas.size } );
+
+        const auto apply_pressure = [&]( double x ) {
+            return units.to_si( UnitSystem::measure::pressure, x );
+        };
+
+        const auto apply_temperature = [=]( double x ) {
+            return units.to_si( UnitSystem::measure::temperature, x );
+        };
+
+        std::transform( sol[ ds::PRESSURE ].begin(), sol[ ds::PRESSURE ].end(),
+                        sol[ ds::PRESSURE ].begin(), apply_pressure );
+        std::transform( sol[ ds::TEMP ].begin(), sol[ ds::TEMP ].end(),
+                        sol[ ds::TEMP ].begin(), apply_temperature );
+
+        /* optional keywords */
```

Comment:
The `RS`and `RV`keywords are not really optional - if the model is specified with the `DISGAS` and `VAPOIL` keywords these keywords *should* be present in the restart file.